cjwkernel/types.py

```diff
@@ -592,16 +592,26 @@ class I18nMessage:
     args: Dict[str, Union[int, float, str]] = field(default_factory=dict)
     """Arguments (empty if message does not need any -- which is common)."""
 
+    source: Dict[str, str] = field(default_factory=dict)
```

Comment:
I was envisioning:

```
@dataclass(frozen=True)
class I18nMessageSource:
    pass

@dataclass(frozen=True)
class I18nMessageSourceLibrary(I18nMessageSource):
    library: str

@dataclass(frozen=True)
class I18nMessageSourceModule(I18nMessageSource):
    module_id: str

I18nMessageSource.Library = I18nMessageSourceLibrary
I18nMessageSource.Module = I18nMessageSourceModule

# ... and then here, within I18nMessage:
    source: Optional[I18nMessageSource]
```

My reasoning behind `Optional`: our database is already chock-full of `I18nMessage` values (all `{"id": "TODO_i18n", {"text": "..."}}`), and in _those_ ones, `source` is undefined. So I'd feel uncomfortable storing `source: {}` in other `I18nMessage` values in the database, because then we'd have two representations of the same thing in the database.

My reasoning behind strongly-typed data classes: that's what this module is all about. (See `ColumnType`, for instance.)